src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -746,6 +746,11 @@ private async ValueTask<int> EnsureFullTlsFrameAsync<TIOAdapter>(TIOAdapter adap
                 return frameSize;
             }
 
+            if (frameSize < int.MaxValue)
+            {
+                _buffer.EnsureAvailableSpace(frameSize - _buffer.EncryptedLength);
+            }
+
```

Comment:
When supplying store-based `SslCertificateTrust`, the ServerHello can grow up substantially (well over 8 KB buffer used until now, so we may need to resize the buffer.

Reply:
I'm not sure if this is needed. If the hello is bigger than maximum frame size it should arrive in two (or more) frames. And that should be OK IMHO. If it is not, we should figure out what is happening. 
This would be problem already, right? e.g. this is unrelated to sending the list since it is in receiving path.

Reply:
There is a way for the code to hang without this:
- Default buffer size during handshake is 8kB
- (one of) the incoming frames is larger than 8kB, the size is known before entering the while loop below the added code
- first read fills in rest of the buffer, part of the frame remains in the inner stream
- since the frame was known, the [call](https://github.com/dotnet/runtime/pull/65195/files/69c523754a7a3e1f48cac345a70afc4e54882d92#diff-1d28e92dd65ace08e351f991ffd5da78f606a6bc4f212160df97d5012bd7cdc9R777) to `_buffer.EnsureAvailableSpace` is skipped
- on second iteration, 0 bytes are read since there is no available space to read into
- `throw new IOException(SR.net_io_eof)`

Reply:
Is there chance this will mess up our buffer logic? I briefly look at the `EnsureAvailableSpace`  implementation and it can for example change _activeStart e.g. move the data within the buffer.  I'm wondering if we should call EnsureAvailableSpace immediately when we determine frame size...?

Reply:
I don't think we can mess things up here. The compacting/moving of the data was happening even before we merged the `_internalBuffer` and `_handshakeBuffer`

> I'm wondering if we should call EnsureAvailableSpace immediately when we determine frame size...?

That is what is happening here, unless I misunderstand you. The frame size is read in the `HaveFullTlsFrame` call above and (if the size wasnt' known at that time) at the end of the while loop below.

The alternative place to call `EnsureAvailableSpace` would be right after we finish processing the previous frame and that does not feel right.